README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# psymonitor <img src="man/figures/logo.png" align="right" height=139/>

[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![Travis build
status](https://travis-ci.org/itamarcaspi/psymonitor.svg?branch=master)](https://travis-ci.org/itamarcaspi/psymonitor)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/itamarcaspi/psymonitor?branch=master&svg=true)](https://ci.appveyor.com/project/itamarcaspi/psymonitor)
[![CRAN
status](https://www.r-pkg.org/badges/version/psymonitor)](https://cran.r-project.org/package=psymonitor)
[![metacran
downloads](https://cranlogs.r-pkg.org/badges/psymonitor)](https://cran.r-project.org/package=psymonitor)

`psymonitor` provides an accessible implementation of the popular
real-time monitoring strategy proposed by Phillips, Shi and Yu
(2015a,b;PSY), along with a new bootstrap procedure designed to mitigate
the potential impact of heteroskedasticity and to effect family-wise
size control in recursive testing algorithms (Phillips and Shi,
forthcoming). This methodology has been shown effective for bubble and
crisis detection (PSY, 2015a,b; Phillips and Shi, 2017) and is now
widely used by academic researchers, central bank economists, and fiscal
regulators.

## Installation

You can install the **stable** version from
[CRAN](https://cran.r-project.org/web/packages/psymonitor/index.html)

``` r
install.packages("psymonitor")
```

You can install the **development** version from
[GitHub](https://github.com/itamarcaspi/psymonitor/)

``` r
# install.packages("devtools")
devtools::install_github("itamarcaspi/psymonitor")
```

## Usage

For the illustration purposes we will use data on the credit risk in the
European sovereign sector, that is proxied by an index constructed as a
GDP weighted 10-year government bond yield of the GIIPS (Greece,
Ireland, Italy, Portugal, and Spain) countries, and comes with the
‘psymonitor’ package.

Let’s walk through some basics. First load the `psymonitor` package and
get data on GIIPS spread.

``` r
library(psymonitor)
data(spread)
```

Next, define a few parameters for the test and the simulation.

``` r
y        <- spread$value
obs      <- length(y)
swindow0 <- floor(obs * (0.01 + 1.8 / sqrt(obs))) # set minimal window size
IC       <- 2  # use BIC to select the number of lags
adflag   <- 6  # set the maximum nuber of lags to 6
yr       <- 2  
Tb       <- 12*yr + swindow0 - 1  # Set the control sample size
nboot    <- 99  # set the number of replications for the bootstrap
```

Next, estimate the PSY test statistic using `PSY()` and its
corresponding bootstrap-based critical values using `cvPSYwmboot()`.

``` r
bsadf          <- PSY(y, swindow0 = swindow0, IC = IC,
                      adflag = adflag)  # estimate the PSY test statistics sequence

quantilesBsadf <- cvPSYwmboot(y, swindow0 = swindow0, IC = IC,
                              adflag = adflag, Tb = Tb, nboot = 99,
                              nCores = 2) # simulate critical values via wild bootstrap. Note that the number of cores is arbitrarily set to 2.
```

Next, identify crisis periods, defined as periods where the test
statistic is above its corresponding critical value, using the
`locate()` function.

``` r
dim          <- obs - swindow0 + 1 
monitorDates <- spread$date[swindow0:obs]
quantile95   <- quantilesBsadf %*% matrix(1, nrow = 1, ncol = dim)
ind95        <- (bsadf > t(quantile95[2, ])) * 1
periods      <- locate(ind95, monitorDates)  # Locate crisis periods
```

Finally, print a table that holds the identified crisis periods with the
help of the `disp()`
function.

``` r
crisisDates <- disp(periods, obs)  #generate table that holds crisis periods
print(crisisDates)
```

|   |      start |        end |
| -: | ---------: | ---------: |
| 1 | 2008-03-01 | 2008-03-01 |
| 2 | 2008-09-01 | 2009-04-01 |
| 3 | 2010-05-01 | 2012-08-01 |

-----

Pleas check the packages’ articles for an elaborated analysis of the
[spreads
data](https://itamarcaspi.github.io/psymonitor/articles/illustrationBONDS.html),
as well as a demonstration using data on the [S\&P 500 price-to-dividend
ratio](https://itamarcaspi.github.io/psymonitor/articles/illustrationSNP.html).

-----

## References

  - Phillips, P. C. B., & Shi, S.(2017). Detecting financial collapse
    and ballooning sovereign risk. Cowles Foundation Discussion Paper
    No. 2110.
  - Phillips, P. C. B., & Shi, S.(forthcoming). Real time monitoring of
    asset markets: Bubbles and crisis. In Hrishikesh D. Vinod and C.R.
    Rao (Eds.), *Handbook of Statistics Volume 41 - Econometrics Using
    R*.
  - Phillips, P. C. B., Shi, S., & Yu, J. (2015a). Testing for multiple
    bubbles: Historical episodes of exuberance and collapse in the S\&P
    500. *International Economic Review*, 56(4), 1034–1078.
  - Phillips, P. C. B., Shi, S., & Yu, J. (2015b). Testing for multiple
    bubbles: Limit Theory for Real-Time Detectors. *International
    Economic Review*, 56(4), 1079–1134.


NEWS.md
# psymonitor 0.0.2

Fixed RNG compatibility issues.

# psymonitor 0.0.1.9000

* Added rendered vignettes.

# psymonitor 0.0.1

* First stable release to CRAN.

# psymonitor 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/ADF.R
#' @title Estimate the ADF statistic
#'
#' @description \code{ADF} calculates the augmented Dickey-Fuller (ADF) test
#'   statistic with lag order set fixed or selected by AIC or BIC.
#'
#' @param y   A vector, The data.
#' @param IC  An integer. 0 for fixed lag (default) order 1 for AIC and 2 for BIC
#'   (default = 0).
#' @param adflag  An integer. Lag order when IC=0; maximum number of
#'   lags when IC>0 (default = 0).
#'
#' @return Numeric, ADF test statistic.
#'
#' @references Said, S. E., & Dickey, D. A. (1984). Testing for Unit Roots in
#'   ARMA Models of Unknown Order. \emph{Biometrika}, 71(1984), 599--607.
#'
#' @export
#'
#'
#' @examples
#' y <- rnorm(100)
#' tstat <- ADF(y,  IC = 0, adflag = 1)

ADF <- function(y, IC=0, adflag=0) {
  T0    <- length(y)
  T1    <- length(y) - 1
  const <- rep(1,T1)

  y1 <- as.matrix(y[1:T1])
  y0 <- as.matrix(y[2:T0])
  dy <- as.matrix(y0 - y1)
  x1 <- as.matrix(cbind(y1,const))

  t   <- T1 - adflag
  dof <- t - adflag - 2

  if (IC > 0) {
    ICC <- matrix(0,nrow = adflag + 1,ncol = 1)
    ADF <- matrix(0,nrow = adflag + 1,ncol = 1)
    for (k in 0:adflag) {
      # Model specification
      xx   <- x1[(k + 1):T1,]  #@-from k+1 to the end (including y1 and x)-@
      dy01 <- as.matrix(dy[(k + 1):T1,]) #@-from k+1 to the end (including dy0)-@

      if (k > 0) {
        x2 <- cbind(xx, matrix(0,nrow = T1 - k,ncol = k))
        for (j in 1:k) {
          x2[,ncol(xx) + j] <- dy[(k + 1 - j):(T1 - j)] #@-including k lag variables of dy in x2-@
        }
      } else x2 <- xx

      # OLS regression
      beta <- solve(t(x2) %*% x2) %*% (t(x2) %*% dy01)   #@-model A-@ ### need work
      eps  <- dy01 - x2 %*% beta
      # Information Criteria
      npdf <- sum(-1/2*log(2*pi) - 1/2*(eps^2))
      if (IC == 1) {       #@ AIC @
        ICC[k + 1,] <- -2*npdf/t + 2*nrow(beta)/t
      } else if (IC == 2) {  #@ BIC @
        ICC[k + 1,] <- -2*npdf/t + nrow(beta)*log(t)/t
      }
      se <- t(eps) %*% eps/dof
      sig <- sqrt(diag(matrix(se,nrow(beta),nrow(beta))*solve(t(x2) %*% x2)))
      ADF[k + 1,] <- beta[1,]/sig[1]
    }
    lag0 <- which.min(ICC)
    ADFlag <- ADF[lag0,]
    #lag=lag0-1;
  } else if (IC == 0) {
    # Model Specification
    xx <- x1[(adflag + 1):T1,]        #@-from k+1 to the end (including y1 and x)-@
    dy01 <- as.matrix(dy[(adflag + 1):T1,]) #@-from k+1 to the end (including dy0)-@

    if (adflag > 0) {
      x2 <- cbind(xx,matrix(0,t,adflag))
      for (j in 1:adflag) {
        x2[,ncol(xx) + j] <- dy[(adflag + 1 - j):(T1 - j)] #@-including k lag variables of dy in x2-@
      }
    } else x2 <- xx

    # OLS regression
    beta <- solve(t(x2) %*% x2) %*% (t(x2) %*% dy01)   #@-model A-@
    eps <- dy01 - x2 %*% beta
    se <- t(eps) %*% eps/dof
    sig <- sqrt(diag(matrix(se,nrow(beta),nrow(beta))*solve(t(x2) %*% x2)))
    ADFlag <- beta[1,]/sig[1]
  }

  if (IC == 0) {
    result <- list('fixed lag of order 1' = ADFlag)
  }
  if (IC == 1) {
    result <- list('ADF Statistic using AIC' = ADFlag)
  }
  if (IC == 2) {
    result <- list('ADF Statistic using BIC' = ADFlag)
  }
  return(result)
}


R/PSY.R
#' @title Estimate PSY's BSADF sequence of test statistics
#'
#' @description \code{PSY} implements the real time bubble detection procedure
#'   of Phillips, Shi and Yu (2015a,b)
#'
#' @param y   A vector. The data.
#' @param swindow0 A positive integer. Minimum window size (default = \eqn{T
#'   (0.01 + 1.8/\sqrt{T})}, where \eqn{T} denotes the sample size)
#' @param IC  An integer. 0 for fixed lag order (default), 1 for AIC and 2 for
#'   BIC (default = 0).
#' @param adflag  An integer, lag order when IC=0; maximum number of
#'   lags when IC>0 (default = 0).
#'
#' @return Vector, BSADF test statistic.
#'
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015a). Testing for
#'   multiple bubbles: Historical episodes of exuberance and collapse in the S&P
#'   500. \emph{International Economic Review}, 56(4), 1034--1078.
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015b). Testing for
#'   multiple bubbles: Limit Theory for Real-Time Detectors. \emph{International
#'   Economic Review}, 56(4), 1079--1134.
#'
#' @export
#'
#'
#' @examples
#'
#' y     <- rnorm(80)
#' bsadf <- PSY(y, IC = 0, adflag = 1)
#'


PSY <- function(y, swindow0, IC=0, adflag=0) {

  t <- length(y)

  if (missing(swindow0)) {
    swindow0 <- floor(t * (0.01 + 1.8 / sqrt(t)))
  }

  bsadfs <- matrix(data = NA, nrow = t, ncol = 1)

  for (r2 in swindow0:t) {
    rwadft <- matrix(data = -999, nrow = r2 - swindow0 + 1, ncol = 1)
    for (r1 in 1:(r2 - swindow0 + 1)) {
      rwadft[r1] <- as.numeric(ADF(y[r1:r2], IC, adflag)) # two tail 5% significant level
    }

    bsadfs[r2, 1] <- max(unlist(rwadft))
  }

  bsadf <- bsadfs[swindow0:t]

  return(bsadf)
}


R/disp.R
#' @title Generate a table with identified bubble/crisis periods
#'
#' @description \code{disp} generates a data frame with bubble/crisis
#'   periods identified by the PSY procedure
#'
#' @param obs   A positive integer. The number of observations.
#' @param OT A date vector. Bubbles/crisis periods identified by the
#'   \code{spymonitor::locate} function.
#'
#' @return A vector of strings with bubble/crisis periods.
#'
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015a). Testing for
#'   multiple bubbles: Historical episodes of exuberance and collapse in the S&P
#'   500. \emph{International Economic Review}, 56(4), 1034--1078.
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015b). Testing for
#'   multiple bubbles: Limit Theory for Real-Time Detectors. \emph{International
#'   Economic Review}, 56(4), 1079--1134.
#' @references * Phillips, P. C. B., & Shi, S.(forthcoming). Real time
#'   monitoring of asset markets: Bubbles and crisis. In Hrishikesh D. Vinod and
#'   C.R. Rao (Eds.), \emph{Handbook of Statistics Volume 41 - Econometrics
#'   Using R}.
#'
#' @export
#'
#'
#' @examples
#' data(spread)
#'
#' y        <- spread$value[150:200]
#' obs      <- length(y)
#' swindow0 <- floor(obs*(0.01 + 1.8/sqrt(obs)))
#' dim      <- obs - swindow0 + 1
#' Tb       <- 24 + swindow0 - 1
#'
#' # Estimate PSY statistics and CVs
#' bsadf          <- PSY(y, swindow0)
#' quantilesBsadf <- cvPSYwmboot(y, swindow0, Tb=Tb, nboot = 49, nCores = 2)
#' quantile95     <- quantilesBsadf %*% matrix(1, nrow = 1, ncol = dim)
#'
#' # locate bubble/crisis dates
#' ind95        <- (bsadf > t(quantile95[2, ])) * 1
#' monitorDates <- spread$date[swindow0:obs]
#' OT           <- locate(ind95, monitorDates)
#'
#' # Show bubble/crisis periods
#' disp(OT, obs)
#'

disp <- function(OT, obs) {
  if (is.null(OT)) {
    stop("No bubble or crisis periods found.", call. = FALSE)
  }

  v <- nrow(OT)
  dateStamps <- data.frame(start = NULL, end = NULL)
  rN <- sample(1:obs, v, replace = TRUE)
  for (j in 1:v) {
    if (OT[j, 1] == OT[j, 2]) {
      newEntry <- data.frame(start = OT[j, 1],
                             end = OT[j, 1])
      dateStamps <- rbind(dateStamps, newEntry)

    } else {
      newEntry <- data.frame(start = OT[j, 1],
                             end = OT[j, 2])
      dateStamps <- rbind(dateStamps, newEntry)

    }
  }
  dateStamps
}


R/ADFres.R
#' @title Estimate the ADF model under the null
#'
#' @description \code{ADFres} estimates the ADF model under the null with lag
#'   order selected by AIC or BIC
#'
#' @param y   A Vector. Data.
#' @param IC  An integer, 0 for fixed lag order (default), 1 for AIC and 2 for
#'   BIC.
#' @param adflag  An integer. Lag order when IC=0; maximum number of lags when
#'   IC>0 (default = 0).
#'
#' @return Numeric, ADF test statistic.
#'
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015a). Testing for
#'   multiple bubbles: Historical episodes of exuberance and collapse in the S&P
#'   500. \emph{International Economic Review}, 56(4), 1034--1078. Phillips, P.
#'   C. B., Shi, S., & Yu, J. (2015b). Testing for multiple bubbles: Limit
#'   Theory for Real-Time Detectors. \emph{International Economic Review},
#'   56(4), 1079--1134.
#'
#'


ADFres <- function(y, IC, adflag) {
  T0    <- length(y)
  T1    <- length(y) - 1
  const <- rep(1,T1)

  dy <- y[2:T0] - y[1:T1]
  x1 <- data.frame(const)

  t <- T1 - adflag
  if (IC > 0) {
    ICC   <- matrix(0,nrow = adflag+1,ncol=1)
    betaM <- matrix(list(), nrow=adflag+1,ncol=1)
    epsM  <- matrix(list(), nrow=adflag+1,ncol=1)

    for (k in 0:adflag){
      # model Specification
      xx<-matrix(x1[(k+1):T1,])  #@-from k+1 to the end (including y1 and x)-@
      dy01<-matrix(dy[(k+1):T1]) #@-from k+1 to the end (including dy0)-@
      if (k>0){
        x2<-cbind(xx,matrix(0,nrow=T1-k,ncol=k))
        for (j in 1:k){
          x2[,ncol(xx)+j]<-dy[(k+1-j):(T1-j)] #@-including k lag variables of dy in x2-@
        }
      }else x2<-xx
      #OLS regression
      betaM[k+1] <- list(solve(t(x2)%*%x2) %*% (t(x2)%*%dy01)) #@-model A-@
      epsM[[k+1]] <- dy01-x2%*%as.matrix(betaM[[k+1]])
      # Information Criteria
      npdf <- sum(-1/2*log(2*pi)-1/2*(epsM[[k+1]]^2))
      if (IC==1){       #@ AIC @
        ICC[k+1] <- -2*npdf/t+2*length(betaM[[k+1]])/t
      }else if(IC==2){ #@ BIC @
        ICC[k+1] <- -2*npdf/t+length(betaM[[k+1]])*log(t)/t
      }
    }
    lag0 <- which.min(ICC)
    beta<-betaM[[lag0]]
    eps<-epsM[[lag0]]
    lag<-lag0-1

  }else if(IC==0){
    # Model Specification
    xx <- matrix(x1[(adflag+1):T1,])    #@-from k+1 to the end (including y1 and x)-@
    dy01 <- matrix(dy[(adflag+1):T1])   #@-from k+1 to the end (including dy0)-@

    if (adflag>0){
      x2<-cbind(xx, matrix(0,nrow=t,ncol=adflag))
      for (j in 1:adflag){
        x2[,ncol(xx)+j]<-dy[(adflag+1-j):(T1-j)] # @-including k lag variables of dy in x2-@
      }
    }else x2 <- xx
    # OLS Regression
    beta <- solve(t(x2)%*%x2) %*% (t(x2)%*%dy01) #@-model A-@
    eps <- dy01-x2%*%beta
    lag<-adflag
  }
  result<-list(beta=beta,eps=eps,lag=lag)
  return(result)
}




R/locate.R
#' @title Locate bubble/crisis periods
#'
#' @description \code{locate} locate bubble/crisis periods with non-zero bubble indicator
#'
#' @param index A vector. A dummy variable that equals 1 for a bubble/crisis
#'   period and 0 otherwise.
#' @param dates A vector. Dates of the time series.
#'
#' @return A vector. Dates identified as bubbles or crisis.
#'
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015a). Testing for
#'   multiple bubbles: Historical episodes of exuberance and collapse in the S&P
#'   500. \emph{International Economic Review}, 56(4), 1034--1078.
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015b). Testing for
#'   multiple bubbles: Limit Theory for Real-Time Detectors. \emph{International
#'   Economic Review}, 56(4), 1079--1134.
#' @references * Phillips, P. C. B., & Shi, S.(forthcoming). Real time
#'   monitoring of asset markets: Bubbles and crisis. In Hrishikesh D. Vinod and
#'   C.R. Rao (Eds.), \emph{Handbook of Statistics Volume 41 - Econometrics
#'   Using R}.
#'
#' @export
#'


locate <- function(index, dates) {
  maxi <- max(index)
  lc <- which.max(index)

  if (maxi == 1) { # there is at least one episode
    count <- 1
    EP <- matrix(0, nrow = 30, ncol = 2)
    # maximum 20 episodes: col1 origination date col2 termination date
    i <- lc + 1
    EP[count, 1] <- dates[lc]
    while (i <= length(index)) {
      if (index[i - 1] == 1 && index[i] == 0) {
        EP[count, 2] <- dates[i - 1]
        i <- i + 1
      } else if (index[i - 1] == 0 && index[i] == 1) {
        count <- count + 1
        EP[count, 1] <- dates[i]
        i <- i + 1
      } else {
        i <- i + 1
      }
    }
    OT <- EP[1:count, ]
    v <- nrow(OT)
    if (OT[v, 2] == 0) {
      OT[v, 2] <- dates[length(dates)]
    }
    OT <- as.Date(OT, origin = "1970-01-01")
  } else if (maxi == 0) {
    OT <- NULL
    warning("No bubble or crisis periods found", call. = FALSE)
  }

  return(OT)
}


R/cvPSYmc.R
#' @title Simulate the finite sample critical values for the PSY test.
#'
#' @description \code{cvPSYmc} implements the real time bubble detection
#'   procedure of Phillips, Shi and Yu (2015a,b)
#'
#' @param obs   A positive integer. The number of observations.
#' @param swindow0 A positive integer. Minimum window size (default = \eqn{T
#'   (0.01 + 1.8/\sqrt{T})}, where \eqn{T} denotes the sample size)
#' @param IC  An integer. 0 for fixed lag order (default), 1 for AIC and 2 for
#'   BIC (default = 0).
#' @param adflag  An integer, lag order when IC=0; maximum number of
#'   lags when IC>0 (default = 0).
#' @param multiplicity Logical. If \code{multiplicity=TRUE}, use family-wise
#'   size control in the recursive testing algorithms.
#' @param Tb A positive integer. The simulated sample size (swindow0+
#'   controlling). Ignored if \code{multiplicity=FALSE}.
#' @param nrep A positive integer. Number of replications (default = 199).
#' @param useParallel Logical. If \code{useParallel=TRUE}, use multi core
#'   computation.
#' @param nCores A positive integer. Optional. If \code{useParallel=TRUE}, the
#'   number of cores defaults to all but one.
#'
#' @return A matrix. BSADF bootstrap critical value sequence at the 90, 95 and
#'   99 percent level.
#'
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015a). Testing for
#'   multiple bubbles: Historical episodes of exuberance and collapse in the S&P
#'   500. \emph{International Economic Review}, 56(4), 1034--1078.
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015b). Testing for
#'   multiple bubbles: Limit Theory for Real-Time Detectors. \emph{International
#'   Economic Review}, 56(4), 1079--1134.
#'
#' @export
#'
#' @import doParallel
#' @import parallel
#' @import foreach
#' @importFrom stats rnorm
#' @importFrom stats quantile
#'
#' @examples
#' cv <- cvPSYmc(80, IC = 0, adflag = 1, Tb = 30, nrep = 99, nCores = 1)

cvPSYmc <- function(obs, swindow0, IC=0, adflag=0, nrep=199,
                    multiplicity=TRUE, Tb, useParallel=TRUE, nCores) {

  qe <- as.matrix(c(0.90, 0.95, 0.99))

  if (missing(swindow0)) {
    swindow0 <- floor(obs * (0.01 + 1.8 / sqrt(obs)))
  }

  if (multiplicity) {
    if (missing(Tb)) {
      stop("Missing a value for 'Tb'", call. = FALSE)
    } else {
      obs <- Tb
    }
  }

  dim <- obs - swindow0 + 1

  suppressWarnings(RNGversion("3.5.0"))

  set.seed(101)
  e <- replicate(nrep, rnorm(obs))
  a <- obs^(-1)
  z <- e + a
  y <- apply(z, 2, cumsum)

  # setup parallel backend
  if (useParallel == TRUE && missing(nCores)) {
    nCores <- detectCores() - 1
  } else {
    nCores <- 1
  }
  cl <- makeCluster(nCores)
  registerDoParallel(cl)

  #----------------------------------
  i <- 0
  MPSY <- foreach(i = 1:nrep, .inorder = FALSE, .combine = rbind) %dopar% {
    PSY(y[, i], swindow0, IC, adflag)
  }
  #----------------------------------

  stopCluster(cl)

  Q_PSY <- as.matrix(quantile(MPSY, qe), na.rm = TRUE)


  return(Q_PSY)
}


R/cvPSYwmboot.R
#' @title Conduct the new composite bootstrapping for the PSY test.
#'
#' @description  \code{cvPSYwmboot} implements the new bootstrap procedure designed
#'   to detect bubbles and crisis periods while mitigating the potential impact
#'   of heteroskedasticity and to effect family-wise size control in recursive
#'   testing algorithms (Phillips and Shi, forthcoming).
#'
#' @param y   A vector. The data.
#' @param swindow0 A positive integer. Minimum window size (default = \eqn{T
#'   (0.01 + 1.8/\sqrt{T})}, where \eqn{T} denotes the sample size),
#' @param IC  An integer. 0 for fixed lag order (default), 1 for AIC and 2 for
#'   BIC (default = 0).
#' @param adflag  An integer, lag order when IC=0; maximum number of
#'   lags when IC>0 (default = 0).
#' @param Tb A positive integer. The simulated sample size (swindow0+
#'   controlling).
#' @param nboot A positive integer. Number of bootstrap replications (default =
#'   199).
#' @param useParallel Logical. If \code{useParallel=TRUE}, use multi core
#'   computation.
#' @param nCores A positive integer. Optional. If \code{useParallel=TRUE}, the
#'   number of cores defaults to all but one.
#'
#' @return A matrix. BSADF bootstrap critical value sequence at the 90, 95 and
#'   99 percent level.
#'
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015a). Testing for
#'   multiple bubbles: Historical episodes of exuberance and collapse in the S&P
#'   500. \emph{International Economic Review}, 56(4), 1034--1078.
#' @references Phillips, P. C. B., Shi, S., & Yu, J. (2015b). Testing for
#'   multiple bubbles: Limit Theory for Real-Time Detectors. \emph{International
#'   Economic Review}, 56(4), 1079--1134.
#' @references Phillips, P. C. B., & Shi, S.(forthcoming). Real time
#'   monitoring of asset markets: Bubbles and crisis. In Hrishikesh D. Vinod and
#'   C.R. Rao (Eds.), \emph{Handbook of Statistics Volume 41 - Econometrics
#'   Using R}.
#'
#' @export
#'
#' @import doParallel
#' @import parallel
#' @import foreach
#'
#' @examples
#' y <- rnorm(80)
#' cv <- cvPSYwmboot(y, IC = 0, adflag = 1, Tb = 30, nboot = 99, nCores = 1)
#'

cvPSYwmboot <- function(y, swindow0, IC=0, adflag=0, Tb, nboot=199,
                   useParallel=TRUE, nCores) {
  qe    <- as.matrix(c(0.90, 0.95, 0.99))
  nboot <- nboot

  result <- ADFres(y, IC, adflag)
  beta   <- result[[1]]
  eps    <- result[[2]]
  lag    <- result[[3]]

  T0 <- length(eps)
  t  <- length(y)
  dy <- as.matrix(y[2:t] - y[1:(t - 1)])
  g  <- length(beta)

  if (missing(swindow0)) {
    swindow0 <- floor(t * (0.01 + 1.8 / sqrt(t)))
  }

  if (missing(Tb)) {
    stop("Missing a value for 'Tb'", call. = FALSE)
  }

  # The DGP
  suppressWarnings(RNGversion("3.5.0"))
  set.seed(101)
  rN <- matrix(sample(1:T0, Tb * nboot, replace = TRUE),
               nrow = Tb, ncol = nboot)
  wn <- matrix(rnorm(1), nrow = Tb, ncol = nboot)

  dyb <- matrix(0, nrow = Tb - 1, ncol = nboot)
  dyb[1:lag, ] <- rep(dy[1:lag], times = nboot)

  for (j in 1:nboot) {
    if (lag == 0) {
      for (i in (lag + 1):(Tb - 1)) {
        dyb[i, j] <- wn[i - lag, j] * eps[rN[i - lag, j]]
      }
    }
    else if (lag > 0) {
      x <- matrix(0, nrow = Tb - 1, ncol = lag)
      for (i in (lag + 1):(Tb - 1)) {
        x <- matrix(0, nrow = Tb - 1, ncol = lag)
        for (k in 1:lag) {
          x[i, (k)] <- dyb[(i - k), j]
        }
        dyb[i, j] <- x[i, ] %*% beta[2:g, 1] + wn[i - lag, j] * eps[rN[i - lag, j]]
      }
    }
  }

  yb0  <- rep(y[1], times = nboot)
  dyb0 <- rbind(yb0, dyb)
  yb   <- apply(dyb0, 2, cumsum)


  # The PSY Test ------------------------------------------------------------

  # setup parallel backend to use many processors
  if (useParallel == TRUE && missing(nCores)) {
    nCores <- detectCores() - 1
  } else {
    nCores <- 1
  }
  cl <- makeCluster(nCores)
  registerDoParallel(cl)

  #----------------------------------
  dim  <- Tb - swindow0 + 1
  i <- 0
  MPSY <- foreach(i = 1:nboot, .inorder = FALSE, .combine = rbind) %dopar% {
    PSY(yb[, i], swindow0, IC, adflag)
  }
  #----------------------------------
  stopCluster(cl)

  SPSY   <- as.matrix(apply(MPSY, 1, max))
  Q_SPSY <- as.matrix(quantile(SPSY, qe))
  return(Q_SPSY)
}
